### ISSUE TITLE
Figure out if we should use globalisel for DXIL lowering

### ISSUE THREAD

**Issue by damyanp** [llvm/wg-hlsl#86]
- [x] Add a cl:opt flag to toggle GlobalIsel on or off
- [x] Add support for an LLVM to MIR translation via the irtranslator pass 
    - Test with `-stop-after=irtranslator`
- [x] Add support for a legalization pass
     - Test with `-stop-after= legalizer ` and `-run-pass=legalizer`
- [x] Add news passes to llvm/lib/Target/DirectX/DirectXTargetMachine.cpp and disable register allocation checks and passes that are needed for physical registers
- [x] Try to use as much of existing coreIsel pass infra as possible
- [x] Add support for an instruction selector
 - Test with `-stop-after= instruction-select ` and `-run-pass=instruction-select`
- [x]  Run through All GlobalIsel passes without stopping
- [ ] convert MIR to DXIL 

https://github.com/llvm/llvm-project/compare/main...farzonl:llvm-project:directxGlobalIsel

**Comment by farzonl** [comment]
The options to convert MIR to DXIL are to  have MIR emit LLVMIR or create a MIR BitcodeWriter.
MIR to LLVMIR has potential perf and memory degrdations. Doubling the LLVMIR memory size in the worst case.
Writing a MIR bitcode writer will need to be able to account for the fact that LLVM IR is value based because of SSA and MIR is not because phsyical registers can't be represented in SSA. For example Instruction operands are values while MIR instruction operands are MachineOperands. Even with Virtual Registers MIR doesn't have values so all the value based data structures would need to be rewritten.

TLDR we have decided not to proceed with GlobalIsel because it would require a rewrite of the DXILBitcodewriter.
